freqtrade/rpc/external_message_consumer.py

```diff
@@ -0,0 +1,251 @@
+            logger.info("Starting ExternalMessageConsumer")
+
+            self._thread = Thread(target=self._loop.run_forever)
+            self._thread.start()
+            self._running = True
+        else:
+            raise RuntimeError("A loop is already running")
+
+        self._main_task = asyncio.run_coroutine_threadsafe(self._main(), loop=self._loop)
+
+    def shutdown(self):
+        """
+        Shutdown the loop, thread, and tasks
+        """
+        if self._thread and self._loop:
+            logger.info("Stopping ExternalMessageConsumer")
+
+            if self._sub_tasks:
+                # Cancel sub tasks
+                for task in self._sub_tasks:
+                    task.cancel()
+
+            if self._main_task:
+                # Cancel the main task
+                self._main_task.cancel()
+
+            self._thread.join()
+
+    async def _main(self):
+        """
+        The main task coroutine
+        """
+        lock = asyncio.Lock()
+
+        try:
+            # Create a connection to each producer
+            self._sub_tasks = [
+                self._loop.create_task(self._handle_producer_connection(producer, lock))
+                for producer in self.producers
+            ]
+
+            await asyncio.gather(*self._sub_tasks)
+        except asyncio.CancelledError:
+            pass
+        finally:
+            # Stop the loop once we are done
+            self._loop.stop()
+
+    async def _handle_producer_connection(self, producer: Dict[str, Any], lock: asyncio.Lock):
+        """
+        Main connection loop for the consumer
+
+        :param producer: Dictionary containing producer info: {'url': '', 'ws_token': ''}
+        :param lock: An asyncio Lock
+        """
+        try:
+            await self._create_connection(producer, lock)
+        except asyncio.CancelledError:
+            # Exit silently
+            pass
+
+    async def _create_connection(self, producer: Dict[str, Any], lock: asyncio.Lock):
+        """
+        Actually creates and handles the websocket connection, pinging on timeout
+        and handling connection errors.
+
+        :param producer: Dictionary containing producer info: {'url': '', 'ws_token': ''}
+        :param lock: An asyncio Lock
+        """
+        while self._running:
+            try:
+                url, token = producer['url'], producer['ws_token']
+                ws_url = f"{url}?token={token}"
+
+                # This will raise InvalidURI if the url is bad
+                async with websockets.connect(ws_url) as ws:
+                    logger.info("Connection successful")
+                    channel = WebSocketChannel(ws)
+
+                    # Tell the producer we only want these topics
+                    # Should always be the first thing we send
+                    await channel.send(
+                        self.compose_consumer_request(RPCRequestType.SUBSCRIBE, self.topics)
+                    )
```

Comment:
the consumer currently has a startup problem (not necessarily tied to this line ...).

Assume for a moment that we'll use "bigger" timeframes (1h and up), and the leader has been running since a day - so data is available at will).

If i now start a follower, it'll send to subscribe to (pairlist, dataframes).
The follower will now wait for data.
In the best case, it's right next to the hour (seconds before :00) - and we get data immediately.
Worst case however - we'll have to wait 59 minutes for the first data to arrive.

I think (for these 2 message types) - we'll have to either
* send the dataframes and whitelist uppon subscription
* allow the subscriber to actively ask for the info through the websocket
* use the rest-api as startup-fallback to avoid a "startup-delay" of up to "1 leader timeframe".

The downside of the first option here is that we're forcing this on all subscribers - even if they maybe received the data through the rest endpoint already.
Using the REST API has the downside that we don't really have auth information for these endpoints - so we'd need to  either allow the API token to all rest endpoints - or provide the follower with user/password as well.

Reply:
I think giving out the username/password like that would be a huge security risk, as we'd only want access to read only data. I think what we could do, is have a special `RPCMessageType.INITIAL_DATA` that we can send when a connection is made, and the consumer could choose to subscribe to it like every other message. Does this work?

Reply:
And then later on, we could rework it to align with requesting data via the websocket if/when that is implemented.